README.md

```diff
@@ -4,19 +4,120 @@ SPDX-FileCopyrightText: 2020-present Open Networking Foundation <info@opennetwor
+Most likely, the exposed operations will be for Read, Write and Stream. The subject of these operations
+should be newly defined entities that will transparently carry various P4Runtime Entries/Updates and
+contain additional status to allow the application to know which state, in the process of applying
+to the data plane, those updates/entities are, e.g., pending, reconciling, applied.
+
+Portions of the stream functionality for packet-out and packet-in are expected to remain without any augmentation.
+It might also be possible to bypass any intervening layers - to be confirmed.
+
+The API would continue to be device centric, i.e., each Read/Write operation would span only a single device.
+Updates given in a single Write should be treated as a transaction, meaning that reconciler should apply all
+derived entities to the device or none of them. There would be no provisions for multi-device operations and transactions.
+
+The API will simplify the mastership arbitration process as much as possible and at the very least,
+will allow the application to learn of the mastership change to allow it to tailor its own activities in response.
+It remains to be seen if the application needs to participate in the mastership selection.
+The library will drive the arbitration process itself.
+
+## Store Controller
+The library should persist these logical pipeline updates/entities to capture the application’s intent
+and allow retrieving it later, either by the API or by the lower levels of the library.
+This functionality is denoted by the Store Controller.
+
+## Translator(s)
+To transform a set of the logical pipeline constructs into another set of device pipeline-specific constructs,
+they need to be subjected to pipeline-specific translation. This layer of the library will manage such
+translation will offer a pipeline-agnostic API to accomplish it. The translator should provide sufficient
+information to allow low-level constructs to be associated with their originating logical pipeline construct.
+This is to appropriately reflect the state of the operation.
+
+To attain run-time extensibility, the translation activities may need to be provided via side-car proxies.
+
+## Reconciliation Controller
+Once a set of pipeline-specific constructs is available, it must be applied to the device.
+This will be the primary objective of the reconciliation controller. In addition to merely applying
+the newly arrived operation to the data-plane, the reconciler also must make sure that the device has
+in fact all the expected entries in its tables. Any departures from the expected state should be addressed
+as soon as possible. The reconciliation approach should consider performance at scale characteristics
+and avoid continuous polling of the state. Instead, it should kick-in only when there is a reasonable
+chance that a departure may have occurred, e.g., extended connection loss or controller outage, port/link outage.
+
+## SB P4Runtime Client
```

Comment:
FYI I am going to add a library for p4runtime client in onos-net-lib as we discussed then we can use it here in SB + for local agents.

Reply:
Precisely.